src/expr/src/relation/func.rs

```diff
@@ -473,7 +473,7 @@ where
     if datums.peek().is_none() {
         Datum::Null
     } else {
-        let x: u64 = datums.map(|d| u64::from(d.unwrap_uint32())).sum();
+        let x: u128 = datums.map(|d| u128::from(d.unwrap_uint32())).sum();
         Datum::from(x)
     }
 }
```

Comment:
Correct me if I'm wrong, but with this PR, we will:
* Error on negative/invalid unsigned accumulations.
* Error on unsigned accumulations that overflows.
* Wrap around on signed accumulations that underflow.
* Wrap around on signed accumulations that overflow.

Is it right that our overflow behavior is different for signed (wrap around) vs unsigned (error) integers?

Reply:
The PR is trying to align the sum behavior under constant folding for `uint2` and `uint4` with the rendering behavior. Previously, we'd accumulate in an `i128` in rendering, then panic if it was outside of the range of a `u64`; however, AFAICT the code would wrap around in a release build under constant folding. So now, we error consistently in both situations.

I am not trying at all to make any changes here to signed behavior, but I'd be happy to discuss what would make sense within the scope of this PR and what would make sense to revisit as follow-up issues. Generally speaking, rendering now will use an `i128` accumulator for all signed integer types. Later, the result is subjected to an `as` conversion to `i64` for sums of `smallint` and `int`, while a `numeric` result is returned for sums of `bigint`. The constant-folding behavior, by contrast, is to use a `sum` over `i64`s for the smaller integer types, which potentially wraps around in release builds. I don't think these two implementations will produce consistent results if the exact same input is represented in, say, a table or as a constant in the query.

Putting aside the differences in constant-folding vs. rendered behavior, we can also ask ourselves if we'd like to have as an outcome that sums wrap around for some types. Not sure what the answer should be here, but I have a feeling that the above might merit a separate issue?

Reply:
I agree, solving this is definitely out of scope for this PR and probably warrants a design doc!

I *think* in general we'd want to do what postgres does. I don't actually know what it does, but if it wraps around, then we should do so too, rather than erroring, IMO. But regardless of what we decide it would be great if we'd wind up with a good understanding of what the correct behavior should be in each case.

Reply:
I created this issue for follow-up work: https://github.com/MaterializeInc/materialize/issues/17758. @jkosh44 may correct me if I'm wrong, but I believe that PostgreSQL will wrap around on signed integer types (https://www.postgresql.org/message-id/3888640A.62BF764E@sferacarta.com) and that it simply does not support unsigned integer types (https://www.postgresql.org/docs/current/datatype-numeric.html) at present.

Reply:
I'm actually not familiar with the PostgreSQL behavior here. Though I agree that we can leave this out of scope for this PR.